### PR TITLE
Feat: support pbr clearcoat

### PR DESCRIPTION
**Pull request by zhuxudong**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)
Feature
* clearcoat
* clearcoatTexture
* clearcoatRoughness
* clearcoatRoughnessTexture
* clearcoatNormalTexture

image

### CODE REVIEW

File: packages/core/src/material/PBRBaseMaterial.ts [hunk 1]
```diff
@@ -21,6 +21,12 @@ export abstract class PBRBaseMaterial extends BaseMaterial {
   private static _emissiveTextureProp = Shader.getPropertyByName("u_emissiveSampler");
   private static _occlusionTextureProp = Shader.getPropertyByName("u_occlusionSampler");
 
+  private static _clearcoatProp = Shader.getPropertyByName("u_clearcoat");
```
Comment (GuoLei1990) [inline packages/core/src/material/PBRBaseMaterial.ts:24]: is `clearCoat` not `clearcoat`

File: packages/core/src/shaderlib/normal_get.glsl [hunk 2]
```diff
@@ -1,47 +1,55 @@
...
+            // get edge vectors of the pixel triangle
+	        vec3 dp1 = dFdx(position);
+	        vec3 dp2 = dFdy(position);
+	        vec2 duv1 = dFdx(uv);
+	        vec2 duv2 = dFdy(uv);
+
+	        // solve the linear system
+	        vec3 dp2perp = cross(dp2, normal);
+	        vec3 dp1perp = cross(normal, dp1);
+	        vec3 tangent = dp2perp * duv1.x + dp1perp * duv2.x;
```
Comment (GuoLei1990) [inline packages/core/src/shaderlib/normal_get.glsl:43]: Maybe Mesh has `tangent`, only no `normal`?

File: packages/core/src/shaderlib/pbr/pbr_helper.glsl [hunk 3]
```diff
@@ -1,44 +1,75 @@
...
-vec3 BRDF_Diffuse_Lambert(vec3 diffuseColor ) {
-	return RECIPROCAL_PI * diffuseColor;
+float getAARoughnessFactor(vec3 normal) {
+    #ifdef HAS_DERIVATIVES
+        vec3 dxy = max( abs(dFdx(normal)), abs(dFdy(normal)) );
+        return max( max(dxy.x, dxy.y), dxy.z );
+    #else
+        return 0.0;
+    #endif
 }
```
Comment (GuoLei1990) [inline packages/core/src/shaderlib/pbr/pbr_helper.glsl:15]: If don't have `HAS_DERIVATIVES` ,not use `0.04` to project?

File: packages/core/src/material/PBRBaseMaterial.ts (hunk 1 above)
Comment (zhuxudong) [reply packages/core/src/material/PBRBaseMaterial.ts:24]: won't fix

Comment (GuoLei1990) [reply packages/core/src/material/PBRBaseMaterial.ts:24]: Won't fix  reason is what? clearcoat will get error by 'Code Speel Checker' and this is not a word.  `Unity` and `babylon`both call this clearCoat not clearcoat.

File: packages/core/src/material/PBRMaterial.ts [hunk 4]
```diff
@@ -23,7 +23,7 @@ export class PBRMaterial extends PBRBaseMaterial {
   }
 
   /**
-   * Roughness.
+   * Roughness, default 1.
```
[the remark this answers is not in the document]
Comment (zhuxudong) [reply packages/core/src/material/PBRMaterial.ts:26]: done

File: packages/core/src/material/PBRBaseMaterial.ts [hunk 5]
```diff
@@ -164,6 +170,85 @@ export abstract class PBRBaseMaterial extends BaseMaterial {
...
+
+  set clearcoat(value: number) {
+    this.shaderData.setFloat(PBRBaseMaterial._clearcoatProp, value);
+
+    if (value === 0) {
+      this.shaderData.disableMacro("CLEARCOAT");
+    } else {
+      this.shaderData.enableMacro("CLEARCOAT");
+    }
+  }
```
[the remark this answers is not in the document]
Comment (zhuxudong) [reply packages/core/src/material/PBRBaseMaterial.ts:188]: done

File: packages/core/src/shaderlib/pbr/pbr_helper.glsl (hunk 3 above)
Comment (zhuxudong) [reply packages/core/src/shaderlib/pbr/pbr_helper.glsl:15]: done

**Pull request merged**